pkg/storegateway/series_chunks.go

```diff
@@ -365,24 +387,61 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 
 			err := c.chunkReaders.addLoad(chunk.blockID, chunk.ref, i, j)
 			if err != nil {
-				c.err = errors.Wrap(err, "preloading chunks")
-				return false
+				return nil, false, errors.Wrap(err, "preloading chunks")
 			}
 		}
 	}
 
 	// Create a batched memory pool that can be released all at once.
 	chunksPool := pool.NewSafeSlabPool[byte](chunkBytesSlicePool, chunkBytesSlabSize)
 
-	err := c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
+	err = c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
 	if err != nil {
-		c.err = errors.Wrap(err, "loading chunks")
-		return false
+		return nil, false, errors.Wrap(err, "loading chunks")
 	}
 
 	nextSet.chunksReleaser = chunksPool
-	c.current = nextSet
-	return true
+
+	if c.ignoreNativeHistograms {
+		nextSet.series = filterNativeHistogramChunks(nextSet.series)
+		if len(nextSet.series) == 0 {
+			nextSet.release()
+			return nil, true, nil
+		}
+	}
+
+	return &nextSet, false, nil
+}
+
+func filterNativeHistogramChunks(seriesEntries []seriesEntry) []seriesEntry {
```

Comment:
Have you benchmarked this? I've the feeling there quite a lot of copying involved when there are no native histograms at all (which is the common case in the initial phase). A simple benchmark may clarify it.

Reply:
We actually have it. Can you run `BenchmarkLoadingSeriesChunksSetIterator` before and after this PR, to compare the impact when native histograms are filtered out but there no native histograms at all in the chunks, please?

Reply:
I had not, but gave it a go:

I made [these changes](https://gist.github.com/leizor/dd4f23d1c51c35123bdc3d6bf19ac75c) for this benchmark (all chunks are native histogram chunks, so all chunks are filtered out) and here are the results:

Current main:
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_10-10         	    2412	    484498 ns/op
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_100-10        	     408	   2922391 ns/op
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_1000-10       	      44	  26052123 ns/op
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_10000-10      	       4	 260503698 ns/op
PASS
ok  	github.com/grafana/mimir/pkg/storegateway	109.621s
```

This branch w/ the above changes:
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_10-10         	    2474	    465029 ns/op
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_100-10        	     417	   2867392 ns/op
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_1000-10       	      46	  25569180 ns/op
BenchmarkLoadingSeriesChunksSetIterator/batch_size:_10000-10      	       4	 252182219 ns/op
PASS
ok  	github.com/grafana/mimir/pkg/storegateway	109.235s
```

Benchstat:
```
name                                                 old time/op  new time/op  delta
LoadingSeriesChunksSetIterator/batch_size:_10-10      484µs ± 0%   465µs ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_100-10    2.92ms ± 0%  2.87ms ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_1000-10   26.1ms ± 0%  25.6ms ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_10000-10   261ms ± 0%   252ms ± 0%   ~     (p=1.000 n=1+1)
```

Reply:
Weird! I thought it'd be at least a bit slower...

Reply:
Can you run the benchmark with `-benchmem` too, please? You can just share the output of `benchstat` for comparison.

Reply:
Yep, absolutely!

```
name                                                 old time/op    new time/op    delta
LoadingSeriesChunksSetIterator/batch_size:_10-10        490µs ± 0%     479µs ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_100-10      2.93ms ± 0%    2.94ms ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_1000-10     25.9ms ± 0%    26.5ms ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_10000-10     257ms ± 0%     262ms ± 0%   ~     (p=1.000 n=1+1)

name                                                 old alloc/op   new alloc/op   delta
LoadingSeriesChunksSetIterator/batch_size:_10-10        158kB ± 0%     133kB ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_100-10       158kB ± 0%     133kB ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_1000-10      189kB ± 0%     164kB ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_10000-10    1.16MB ± 0%    1.13MB ± 0%   ~     (p=1.000 n=1+1)

name                                                 old allocs/op  new allocs/op  delta
LoadingSeriesChunksSetIterator/batch_size:_10-10        2.30k ± 0%     1.40k ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_100-10       2.30k ± 0%     1.40k ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_1000-10      2.71k ± 0%     1.80k ± 0%   ~     (p=1.000 n=1+1)
LoadingSeriesChunksSetIterator/batch_size:_10000-10     3.06k ± 0%     2.16k ± 0%   ~     (p=1.000 n=1+1)
```